File: applications/luci-app-statistics/luasrc/model/cbi/luci_statistics/disk.lua
--[[

Luci configuration model for statistics - collectd disk plugin configuration
(c) 2008 Freifunk Leipzig / Jo-Philipp Wich <xm@leipzig.freifunk.net>

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

        http://www.apache.org/licenses/LICENSE-2.0

$Id$

]]--

m = Map("luci_statistics",
	translate("Disk Plugin Configuration"),
	translate(
		"The disk plugin collects detailled usage statistics " ..
		"for selected partitions or whole disks."
	))

-- collectd_disk config section
s = m:section( NamedSection, "collectd_disk", "luci_statistics" )

-- collectd_disk.enable
enable = s:option( Flag, "enable", translate("Enable this plugin") )
enable.default = 0

-- collectd_disk.disks (Disk)
devices = s:option( Value, "Disks", translate("Monitor disks and partitions") )
devices.default = "hda1 hdb"
devices.rmempty = true
devices:depends( "enable", 1 )

-- collectd_disk.ignoreselected (IgnoreSelected)
ignoreselected = s:option( Flag, "IgnoreSelected", translate("Monitor all except specified") )
ignoreselected.default = 0
ignoreselected:depends( "enable", 1 )

return m

File: applications/luci-app-statistics/htdocs/luci-static/resources/view/statistics/plugins/disk.js
'use strict';
'require baseclass';
'require fs';
'require form';

return baseclass.extend({
	title: _('Disk Plugin Configuration'),
	description: _('The disk plugin collects detailed usage statistics for selected partitions or whole disks.'),

	addFormOptions(s) {
		let o;

		o = s.option(form.Flag, 'enable', _('Enable this plugin'));

		o = s.option(form.DynamicList, 'Disks', _('Monitor disks and partitions'),
			_('When none selected, all disks will be monitored.'));
		o.depends('enable', '1');
		o.load = function(section_id) {
			return fs.trimmed('/proc/partitions').then(L.bind(function(str) {
				const lines = (str || '').split(/\n/);

				for (let line of lines) {
					const m = line.match(/^ +[0-9]+ +[0-9]+ +[0-9]+ (\S+)$/);
					if (m)
						this.value(m[1]);
				}

				return this.super('load', [section_id]);
			}, this));
		};

		o = s.option(form.Flag, 'IgnoreSelected', _('Monitor all except specified'));
		o.depends('enable', '1');
	},

	configSummary(section) {
		const disks = L.toArray(section.Disks);
		const invert = section.IgnoreSelected == '1';

		if (disks.length == 0)
			return _('Monitoring all disks');
		else if (invert)
			return N_(disks.length, 'Monitoring all but one disk', 'Monitoring all but %d disks').format(disks.length);
		else
			return N_(disks.length, 'Monitoring one disk', 'Monitoring %d disks').format(disks.length);
	}
});
